Source/WebCore/loader/NavigationController.cpp

```diff
@@ -0,0 +1,111 @@
+ *     documentation and/or other materials provided with the distribution. 
+ * 3.  Neither the name of Apple Inc. ("Apple") nor the names of
+ *     its contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission. 
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE AND ITS CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL APPLE OR ITS CONTRIBUTORS BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "NavigationController.h"
+
+#include "EventNames.h"
+#include "JavaScriptCore/JSCJSValue.h"
+#include "LocalFrame.h"
+#include "LocalFrameLoaderClient.h"
+#include "NavigateEvent.h"
+#include "Navigation.h"
+#include "NavigationDestination.h"
+#include "NavigationNavigationType.h"
+#include "ResourceRequest.h"
+#include <optional>
+#include <wtf/UUID.h>
+
+namespace WebCore {
+
+FrameLoader::NavigationController::NavigationController(LocalFrame& frame)
+    : m_frame(frame)
+{
+}
+
+FrameLoader::NavigationController::~NavigationController() = default;
+
+void FrameLoader::NavigationController::dispatchNavigationEvent(const RefPtr<NavigationEntry> entry, const URL& oldURL, const Document& document, NavigationNavigationType type)
+{
+    URL newURL { entry->url() };
+    bool isSameDocument = equalIgnoringFragmentIdentifier(newURL, oldURL);
+    bool isHashChange = isSameDocument && !equalRespectingNullity(newURL.fragmentIdentifier(), oldURL.fragmentIdentifier());
+    bool isTraversal = type == NavigationNavigationType::Traverse;
+    size_t index = std::distance(m_navigations.begin(), m_navigations.find(entry->key()));
+
+    auto destination = NavigationDestination::create(*document.domWindow(), entry->url(), isTraversal ? entry->id() : ""_s, isTraversal ? entry->key() : ""_s, isTraversal ? index : -1);
+    auto init = NavigateEvent::Init {
+        { false, true, false },
+        type,
+        WTFMove(destination),
+        nullptr,
+        nullptr,
+        ""_s,
+        JSC::jsUndefined(),
+        true,
+        false,
+        isHashChange,
+        false,
+    };
+    auto event = NavigateEvent::create({ "navigate"_s }, init);
+    m_frame.window()->navigation().dispatchEvent(event);
+}
+
+std::tuple<const RefPtr<NavigationEntry>, int64_t> FrameLoader::NavigationController::currentEntry() const
```

Comment:
Since we can never have one without the other and there are only 2, could this be a std::optional<std::pair<Ref<NavigationEntry>, size_t>> instead?